Dominant Speaker Event

SteveMcFarlin opened the pull request:
Implementation of Dominant Speaker Identification for Multipoint Videoconferencing by Ilana Volfin and Israel Cohen. This implementation uses the RTP Audio Level extension from RFC-6464 for the input signal. This has been ported from DominantSpeakerIdentification.java in [Jitsi](https://github.com/jitsi/jitsi-utils/blob/master/src/main/java/org/jitsi/utils/dsi/DominantSpeakerIdentification.java)

Commit:
Dominant Speaker Event (#5)

Implementation of Dominant Speaker Identification for Multipoint Videoconferencing by Ilana Volfin and Israel Cohen. This implementation uses the RTP Audio Level extension from RFC-6464 for the input signal.

Commit:
Adding attribution.

ibc commented on src/ActiveSpeakerObserver.ts:
```diff
@@ -0,0 +1,66 @@
+import { Producer } from './Producer';
+
+export interface ActiveSpeakerObserverOptions {
+
```
```suggestion
```

ibc commented on src/ActiveSpeakerObserver.ts:
```diff
@@ -0,0 +1,66 @@
+
+export interface ActiveSpeakerObserverActivity {
+	/**
+	 * The producer instance
```
```suggestion
	 * The producer instance.
```

ibc commented on src/ActiveSpeakerObserver.ts:
```diff
@@ -0,0 +1,66 @@
+	 */
+	constructor(params: any)
+	{
+		super(params);
+		this._handleWorkerNotifications();
```
```suggestion
		super(params);

		this._handleWorkerNotifications();
```

ibc commented on src/ActiveSpeakerObserver.ts:
```diff
@@ -0,0 +1,66 @@
+			{
+				case 'dominantSpeaker':
+				{
+					const dominantSpeaker = {
+						producer : this._getProducerById(data.producerId)
+					};
+
+					this.safeEmit('dominantSpeaker', dominantSpeaker);
+					this._observer.safeEmit('dominantSpeaker');
+					break;
+				}
```
```suggestion
					const dominantSpeaker =
					{
						producer : this._getProducerById(data.producerId)
					};

					this.safeEmit('dominantSpeaker', dominantSpeaker);
					this._observer.safeEmit('dominantSpeaker');

					break;
				}

```

ibc commented on src/Router.ts:
```diff
@@ -924,6 +925,47 @@ export class Router extends EnhancedEventEmitter
+		}: ActiveSpeakerObserverOptions = {}
+	): Promise<ActiveSpeakerObserver>
+	{
+		logger.debug('createActiveSpeakerObserver()');
+		if (appData && typeof appData !== 'object')
+			throw new TypeError('if given, appData must be an object');
```
Please let's keep coding conventions everywhere.

```suggestion
		logger.debug('createActiveSpeakerObserver()');

		if (appData && typeof appData !== 'object')
			throw new TypeError('if given, appData must be an object');
```

ibc commented on src/Router.ts:
```diff
@@ -924,6 +925,47 @@ export class Router extends EnhancedEventEmitter
+			this._rtpObservers.delete(activeSpeakerObserver.id);
+		});
+
+		this._observer.safeEmit('newrtpobserver', activeSpeakerObserver);
```
```suggestion
		// Emit observer event.
		this._observer.safeEmit('newrtpobserver', activeSpeakerObserver);
```

ibc commented on worker/compile_commands_template.json:
```diff
@@ -4299,6 +4299,49 @@
         "file": "../src/RTC/Consumer.cpp"
     },
+    {
+        "arguments": [
```
Why this change in this file?

SteveMcFarlin replied:
I assumed I needed to add this in order to compile ActiveSpeakerObserver.cpp. If there is a cleaner way please let me know.

ibc replied:
See docs/Building.md. That's file is generated automatically :)

nazar-pc replied:
This seems like a file for IDE support, but definitely doesn't participate in build.

SteveMcFarlin replied:
Thanks. I will check that and issue a separate PR, after the batch commit.

ibc replied:
Actually no need for that (unless you want to make `make tidy` command work in your host :)

SteveMcFarlin replied:
I meant to say another commit. What would you like me to do for this? I tried running **make bear**, but get the following error: 

`bear -o compile_commands_template.json /Applications/Xcode.app/Contents/Developer/usr/bin/make
Unrecognized parameter: "-o"`

SteveMcFarlin replied:
I tried **--output** with no success. I will see what is going on. Perhaps a version mismatch for bear on OS-X.

SteveMcFarlin replied:
bear 2.1.5 on Big Sur is failing to compile. I will look into it later. I'll address the other issues for now.

SteveMcFarlin replied:
I have bear 2.7 running and working. However, when running tidy I am getting a lot of warnings and errors. The error is it can not find wchar.h. 

`27547 warnings and 1 error generated.`

The resulting compile_commands.json looks correct, and is correctly formatted. I have committed the unformatted compile_commands_template.json for now.

ibc commented on worker/include/RTC/ActiveSpeakerObserver.hpp:
```diff
@@ -0,0 +1,63 @@
+// https://github.com/jitsi/jitsi-utils/blob/master/src/main/java/org/jitsi/utils/dsi/DominantSpeakerIdentification.java
+namespace RTC
+{
+	class Speaker;
```
Please let's define this `Speaker` class within the `ActiveSpeakerObserver` class. We don't want that this header file exposes a global `RTC::Speaker` class.

ibc commented on worker/src/RTC/ActiveSpeakerObserver.cpp:
```diff
@@ -0,0 +1,549 @@
+
+	void ActiveSpeakerObserver::OnTimer(Timer* timer)
+	{
+		MS_TRACE();
+		Update();
```
```suggestion
		MS_TRACE();

		Update();
```

ibc commented on src/ActiveSpeakerObserver.ts:
```diff
@@ -0,0 +1,66 @@
+						producer : this._getProducerById(data.producerId)
+					};
+
+					this.safeEmit('dominantSpeaker', dominantSpeaker);
```
Please use lowcase in JS event names as convention.

```suggestion
					this.safeEmit('dominantspeaker', dominantSpeaker);
```

NOTE: This should be changed in tests as well obviously.

ibc commented on src/ActiveSpeakerObserver.ts:
```diff
@@ -0,0 +1,66 @@
+					};
+
+					this.safeEmit('dominantSpeaker', dominantSpeaker);
+					this._observer.safeEmit('dominantSpeaker');
```
```suggestion
					this._observer.safeEmit('dominantspeaker');
```

ibc commented on src/ActiveSpeakerObserver.ts:
```diff
@@ -0,0 +1,66 @@
+		{
+			switch (event)
+			{
+				case 'dominantSpeaker':
+				{
+					const dominantSpeaker = {
```
No camelcase events in mediasoup.

```suggestion
				case 'dominantspeaker':
				{
					const dominantSpeaker = {
```

ibc reviewed (changes requested):
Amazing, thanks. First review round done.

nazar-pc commented:
It would be great to include corresponding changes on Rust side as well :pray:

Commit:
Apply suggestions from code review

Co-authored-by: Iñaki Baz Castillo <ibc@aliax.net>

Commit:
Apply suggestions from code review

Co-authored-by: Iñaki Baz Castillo <ibc@aliax.net>

Commit:
Revert changes to compile_commands_template.json

Commit:
Moving Speaker class into ActiveSpeakerObserver

SteveMcFarlin commented:
> It would be great to include corresponding changes on Rust side as well 🙏

I took a look at the Rust implementation. I am slightly confused. In the implementation for audio_level_observer.rs it appears the RtpObserver is also implemented in that file. In any case I thought perhaps I could "hack" it if the code was not much. Basically by following the core audio_level_observer.rs implementation. Well I thought wrong. I have only ever played with Rust, so I don't think I can help here.

nazar-pc commented:
> In the implementation for audio_level_observer.rs it appears the RtpObserver is also implemented in that file.

It is not implementation of `RtpObserver` itself, it just implements `RtpObserver` trait for `AudioLevelObserver` struct. Think of it as implementing an interface for a class in other languages.

ibc commented on worker/src/RTC/Router.cpp:
```diff
@@ -270,6 +271,25 @@ namespace RTC
+			{
+				std::string rtpObserverId;
+
+				// This may throw
```
Missing dot :)

ibc commented on src/ActiveSpeakerObserver.ts:
```diff
@@ -0,0 +1,66 @@
+
+					this.safeEmit('dominantspeaker', dominantSpeaker);
+					this._observer.safeEmit('dominantspeaker');
+					break;
```
Missing blank line before break.

Commit:
Fix formatting

jmillan commented on worker/src/Channel/ChannelNotifier.cpp:
```diff
@@ -44,7 +44,6 @@ namespace Channel
 		jsonNotification["targetId"] = targetId;
 		jsonNotification["event"]    = event;
 		jsonNotification["data"]     = data;
-
```
This blank line was OK there.

jmillan commented on worker/src/RTC/ActiveSpeakerObserver.cpp:
```diff
@@ -0,0 +1,523 @@
+	{
+		MS_TRACE();
+
+		if (producer->GetKind() != RTC::Media::Kind::AUDIO)
+			MS_THROW_TYPE_ERROR("not an audio Producer");
+
+		this->mapProducerSpeaker[producer->id].speaker->paused = false;
```
This method is called by Router for every Producer that is resumed. We cannot throw here. Let's just resume the Speaker if the corresponding Producer is in the map.

```suggestion
		auto it = this->mapProducerSpeaker.find(producer->id);
		
		if (it != this->mapProducerSpeaker.end())
		{
			auto& rtpObserver = it->second;
			
			rtpObserver.speaker->paused = false;
		}
```

jmillan commented on worker/src/RTC/ActiveSpeakerObserver.cpp:
```diff
@@ -0,0 +1,523 @@
+		this->mapProducerSpeaker[producer->id].speaker->paused = false;
+	}
+
+	void ActiveSpeakerObserver::ProducerPaused(RTC::Producer* producer)
```
Same as in previous comment, this will crash because the method is called by Router for every paused Producer regardless it's present in `mapProducerSpeaker` map.

jmillan commented on worker/src/RTC/ActiveSpeakerObserver.cpp:
```diff
@@ -0,0 +1,523 @@
+			return;
+
+		uint64_t now = DepLibUV::GetTimeMs();
+		this->mapProducerSpeaker[producer->id].speaker->LevelChanged(volume, now);
```
What if a packet is received from a Producer that is not present in this map? This will crash.

jmillan commented on worker/src/RTC/ActiveSpeakerObserver.cpp:
```diff
@@ -0,0 +1,523 @@
+		{
+			json data          = json::object();
+			data["producerId"] = this->dominantId;
+			std::string s      = data.dump();
```
not used

ggarber committed:
Add rust module support for active speaker observer

Commit:
Rust ActiveSpeakerObserver implementation
Merge remote-tracking branch 'ggb/steve-v3-dominant-speaker' into v3-dominant-speaker

SteveMcFarlin commented:
@nazar-pc Please take a look at the rust implementation. Thanks!

Commit:
Update worker/src/RTC/ActiveSpeakerObserver.cpp

Co-authored-by: José Luis Millán <jmillan@aliax.net>

Commit:
Guard against producer that is not present in map.

Commit:
Remove unused temporary.

SteveMcFarlin commented on worker/src/RTC/ActiveSpeakerObserver.cpp:
```diff
@@ -181,8 +188,15 @@ namespace RTC
+			auto& rtpObserver = it->second;
+			uint64_t now      = DepLibUV::GetTimeMs();
+
+			rtpObserver.speaker->paused = LevelChanged(volume, now);
```
🥇 Will fix this one after testing.

Commit:
Fix code

nazar-pc commented:
Rust version looks good, thank you!

ibc commented on src/ActiveSpeakerObserver.ts:
```diff
@@ -0,0 +1,67 @@
+					this._observer.safeEmit('dominantspeaker');
+
+					break;
+				}
+				default:
```
```suggestion
				}

				default:
```

ibc replied:
Let me commit this one :)

ibc committed:
Update src/ActiveSpeakerObserver.ts

ibc commented:
CI is failing due to this:

```
../include/RTC/ActiveSpeakerObserver.hpp:6:10: fatal error: json.hpp: No such file or directory
    6 | #include <json.hpp>
      |          ^~~~~~~~~~
```

It must be `#include <nlohmann/json.hpp>`.

ibc commented on src/ActiveSpeakerObserver.ts:
```diff
@@ -0,0 +1,68 @@
+					};
+
+					this.safeEmit('dominantspeaker', dominantSpeaker);
+					this._observer.safeEmit('dominantspeaker');
```
```suggestion
					this._observer.safeEmit('dominantspeaker', dominantSpeaker);
```

ibc replied:
Easy one.

ibc committed:
Update src/ActiveSpeakerObserver.ts

ibc commented on worker/src/RTC/ActiveSpeakerObserver.cpp:
```diff
@@ -0,0 +1,540 @@
+			dominantSpeaker->EvalActivityScores();
+			double newDominantC2 = C2;
+
+			for (auto itt = mapProducerSpeaker.begin(); itt != mapProducerSpeaker.end(); itt++)
```
Can we rename `itt` to just `it`?

ibc commented on worker/src/RTC/ActiveSpeakerObserver.cpp:
```diff
@@ -0,0 +1,540 @@
+			default:
+				MS_ASSERT(false, "Interval is invalid");
+				break;
+		}
+		return 0;
```
```suggestion
		}

		return 0;
```

ibc commented on worker/src/RTC/ActiveSpeakerObserver.cpp:
```diff
@@ -0,0 +1,540 @@
+			case 2:
+				return longActivityScore;
+			default:
+				MS_ASSERT(false, "Interval is invalid");
+				break;
```
```suggestion
				MS_ABORT("interval is invalid");
```

ibc committed:
Update worker/src/RTC/ActiveSpeakerObserver.cpp

ibc commented on worker/src/RTC/ActiveSpeakerObserver.cpp:
```diff
@@ -0,0 +1,540 @@
+
+	bool ActiveSpeakerObserver::Speaker::ComputeMediums()
+	{
+		MS_TRACE();
+		return ComputeBigs(immediates, mediums, MediumThreshold);
```
```suggestion
		MS_TRACE();

		return ComputeBigs(immediates, mediums, MediumThreshold);
```

ibc commented on worker/src/RTC/ActiveSpeakerObserver.cpp:
```diff
@@ -0,0 +1,540 @@
+
+	bool ActiveSpeakerObserver::Speaker::ComputeLongs()
+	{
+		MS_TRACE();
+		return ComputeBigs(mediums, longs, LongThreashold);
```
```suggestion
		MS_TRACE();

		return ComputeBigs(mediums, longs, LongThreashold);
```

ibc commented on worker/src/RTC/ActiveSpeakerObserver.cpp:
```diff
@@ -0,0 +1,540 @@
+
+	void ActiveSpeakerObserver::Speaker::EvalImmediateActivityScore()
+	{
+		MS_TRACE();
+		immediateActivityScore = ComputeActivityScore(immediates[0], N1, 0.5, 0.78);
```
```suggestion
		MS_TRACE();

		immediateActivityScore = ComputeActivityScore(immediates[0], N1, 0.5, 0.78);
```

ibc commented on worker/src/RTC/ActiveSpeakerObserver.cpp:
```diff
@@ -0,0 +1,540 @@
+
+	void ActiveSpeakerObserver::Speaker::EvalMediumActivityScore()
+	{
+		MS_TRACE();
+		mediumActivityScore = ComputeActivityScore(mediums[0], N2, 0.5, 24);
```
```suggestion
		MS_TRACE();

		mediumActivityScore = ComputeActivityScore(mediums[0], N2, 0.5, 24);
```

ibc commented on worker/src/RTC/ActiveSpeakerObserver.cpp:
```diff
@@ -0,0 +1,540 @@
+
+	void ActiveSpeakerObserver::Speaker::EvalLongActivityScore()
+	{
+		MS_TRACE();
+		longActivityScore = ComputeActivityScore(longs[0], N3, 0.5, 47);
```
```suggestion
		MS_TRACE();

		longActivityScore = ComputeActivityScore(longs[0], N3, 0.5, 47);
```

ibc commented on worker/src/RTC/ActiveSpeakerObserver.cpp:
```diff
@@ -0,0 +1,539 @@
+		}
+		else if (speakerCount == 1)
+		{
+			auto ittr     = mapProducerSpeaker.begin();
+			newDominantId = ittr->second.producer->id;
```
```suggestion
			auto it = mapProducerSpeaker.begin();

			newDominantId = it->second.producer->id;
```

ibc committed:
Update worker/src/RTC/ActiveSpeakerObserver.cpp

ibc commented on worker/src/RTC/ActiveSpeakerObserver.cpp:
```diff
@@ -0,0 +1,539 @@
+
+		if (!newDominantId.empty() && newDominantId != dominantId)
+		{
+			dominantId = newDominantId;
+			return true;
```
```suggestion
			dominantId = newDominantId;

			return true;
```

ibc commented on worker/src/RTC/ActiveSpeakerObserver.cpp:
```diff
@@ -0,0 +1,539 @@
+			dominantSpeaker->EvalActivityScores();
+			double newDominantC2 = C2;
+
+			for (auto itt = mapProducerSpeaker.begin(); itt != mapProducerSpeaker.end(); itt++)
+			{
+				Speaker* speaker     = itt->second.speaker;
+				const std::string id = itt->second.producer->id;
```
```suggestion
			for (auto it = mapProducerSpeaker.begin(); it != mapProducerSpeaker.end(); ++it)
			{
				Speaker* speaker     = it->second.speaker;
				const std::string id = it->second.producer->id;
```

ibc committed:
Update worker/src/RTC/ActiveSpeakerObserver.cpp

ibc commented on worker/include/RTC/ActiveSpeakerObserver.hpp:
```diff
@@ -0,0 +1,96 @@
+		private:
+			uint8_t minLevel;
+			uint8_t nextMinLevel;
+			uint32_t nextMinLevelWindowLen{ 0 };
+
```
```suggestion
			uint32_t nextMinLevelWindowLen{ 0 };
```

ibc commented on worker/src/RTC/ActiveSpeakerObserver.cpp:
```diff
@@ -0,0 +1,540 @@
+		}
+
+		int64_t t = 1;
+		for (int64_t i = n, j = 1; i > r; i--, j++)
```
```suggestion
		for (int64_t i = n, j = 1; i > r; i--, ++j)
```

ibc commented on worker/src/RTC/ActiveSpeakerObserver.cpp:
```diff
@@ -0,0 +1,540 @@
+		{
+			uint8_t sum = 0;
+
+			for (uint32_t lEnd = l + littleLenPerBig; l < lEnd; l++)
```
```suggestion
			for (uint32_t lEnd = l + littleLenPerBig; l < lEnd; ++l)
```

ibc commented on worker/src/RTC/ActiveSpeakerObserver.cpp:
```diff
@@ -0,0 +1,540 @@
+			{
+				if (littles[l] > threashold)
+				{
+					sum++;
```
```suggestion
					++sum;
```

ibc commented on worker/src/RTC/ActiveSpeakerObserver.cpp:
```diff
@@ -0,0 +1,540 @@
+				{
+					sum++;
+				}
+			}
+			if (bigs[b] != sum)
```
```suggestion
			}

			if (bigs[b] != sum)
```

ibc commented on worker/src/RTC/ActiveSpeakerObserver.cpp:
```diff
@@ -0,0 +1,540 @@
+				bigs[b] = sum;
+				changed = true;
+			}
+		}
+		return changed;
```
```suggestion
		}

		return changed;
```

ibc commented on worker/src/RTC/ActiveSpeakerObserver.cpp:
```diff
@@ -0,0 +1,540 @@
+	{
+		MS_TRACE();
+
+		if (mapProducerSpeaker[producer->id].speaker != nullptr)
```
This is wrong. If `mapProducerSpeaker[producer->id]` does not exist in the map then you are filling it right now here. Please use `.find()` instead.

ibc commented on worker/src/RTC/ActiveSpeakerObserver.cpp:
```diff
@@ -0,0 +1,540 @@
+		MS_TRACE();
+
+		uint64_t now              = DepLibUV::GetTimeMs();
+		uint64_t levelIdleTimeout = LevelIdleTimeout - (now - lastLevelIdleTime);
+		if (levelIdleTimeout <= 0)
```
```suggestion
		uint64_t levelIdleTimeout = LevelIdleTimeout - (now - lastLevelIdleTime);

		if (levelIdleTimeout <= 0)
```

ibc commented on worker/src/RTC/ActiveSpeakerObserver.cpp:
```diff
@@ -0,0 +1,540 @@
+		if (mapProducerSpeaker.size() > 0 && CalculateActiveSpeaker())
+		{
+			json data          = json::object();
+			data["producerId"] = this->dominantId;
+			Channel::ChannelNotifier::Emit(this->id, "dominantSpeaker", data);
```
Rename `dominantSpeaker` to `dominantspeaker`. In fact this was already changed in JS side.

```suggestion
			data["producerId"] = this->dominantId;

			Channel::ChannelNotifier::Emit(this->id, "dominantspeaker", data);
```

ibc commented on worker/src/RTC/ActiveSpeakerObserver.cpp:
```diff
@@ -0,0 +1,540 @@
+	{
+		MS_TRACE();
+
+		for (auto itt = mapProducerSpeaker.begin(); itt != mapProducerSpeaker.end(); itt++)
+		{
+			Speaker* speaker      = itt->second.speaker;
+			const std::string& id = itt->second.producer->id;
+			uint64_t idle         = now - speaker->lastLevelChangeTime;
```
```suggestion
		for (auto it = mapProducerSpeaker.begin(); it != mapProducerSpeaker.end(); ++it)
		{
			Speaker* speaker      = it->second.speaker;
			const std::string& id = it->second.producer->id;
			uint64_t idle         = now - speaker->lastLevelChangeTime;
```

ibc commented on worker/src/RTC/ActiveSpeakerObserver.cpp:
```diff
@@ -0,0 +1,540 @@
+	{
+		MS_TRACE();
+
+		if (ComputeImmediates())
+		{
+			EvalImmediateActivityScore();
+			if (ComputeMediums())
+			{
+				EvalMediumActivityScore();
+				if (ComputeLongs())
+				{
```
```suggestion
			EvalImmediateActivityScore();

			if (ComputeMediums())
			{
				EvalMediumActivityScore();

				if (ComputeLongs())
				{
```

ibc committed:
Apply suggestions from code review

ibc reviewed (changes requested):
I've pushed some minor tweaks. In addition to pending requested changes, an important thing here:

Please use `this->xxxx` instead of just `xxxx` for class private members (and don't use `this->` for class methods). We follow that convention everywhere.

ibc committed:
Update worker/src/RTC/ActiveSpeakerObserver.cpp

ibc committed:
Update worker/src/RTC/ActiveSpeakerObserver.cpp

ibc committed:
Update worker/src/RTC/ActiveSpeakerObserver.cpp

ibc committed:
Update worker/src/RTC/ActiveSpeakerObserver.cpp

ibc committed:
Update worker/src/RTC/ActiveSpeakerObserver.cpp

ibc committed:
Update worker/src/RTC/ActiveSpeakerObserver.cpp

ibc committed:
Update worker/src/RTC/ActiveSpeakerObserver.cpp

ibc committed:
Update worker/src/RTC/ActiveSpeakerObserver.cpp

ibc committed:
Update worker/src/RTC/ActiveSpeakerObserver.cpp

ibc committed:
Update worker/src/RTC/ActiveSpeakerObserver.cpp

ibc committed:
Update worker/src/RTC/ActiveSpeakerObserver.cpp

ibc committed:
Update worker/src/RTC/ActiveSpeakerObserver.cpp

Commit:
Guard against missing map item.

Commit:
Use iterator to avoid another lookup

Commit:
Rebuild JS and TS map files.

Commit:
Rebuild router JS and map file.

ibc commented on src/ActiveSpeakerObserver.ts:
```diff
@@ -0,0 +1,68 @@
+	/**
+	 * The producer instance.
+	 */
+	producer: Producer;
```
Just a question: will the ActiveSpeakerObserver C++ logic always choose a Producer even if nobody is talking? Imagine that eventually we remove all audio Producers from the ActiveSpeakerObserver. Will it emit some event telling `producer: undefined`? Or will it just NOT emit any event?

SteveMcFarlin replied:
The first time through the calculation it will pick the first producer/speaker in the map. If the speaker count is 0 then no event is emitted. This is determined in `ActiveSpeakerObserver::CalculateActiveSpeaker()`.

ibc commented on worker/include/RTC/ActiveSpeakerObserver.hpp:
```diff
@@ -0,0 +1,95 @@
+	private:
+		class Speaker
+		{
+		public:
```
Please let reorder this as we do in other C++ classes:

1. Public static methods first.
2. Then private static methods.
3. Then public static members.
4. Then private static members.
5. Public class methods first.
6. Then private class methods.
7. Then public class members.
8. Then private class members.

This is:

```c++
public:
	Speaker();
	void EvalActivityScores();
	double GetActivityScore(int32_t interval);
	void LevelChanged(uint32_t level, uint64_t time);
	void LevelTimedOut();

private:
	bool ComputeImmediates();
	bool ComputeLongs();
	bool ComputeMediums();
	void EvalImmediateActivityScore();
	void EvalMediumActivityScore();
	void EvalLongActivityScore();
	void UpdateMinLevel(int8_t level);

public:
	bool paused{ false };
	double immediateActivityScore;
	double mediumActivityScore;
	double longActivityScore;
	uint64_t lastLevelChangeTime{ 0 };

private:
	uint8_t minLevel;
	uint8_t nextMinLevel;
	uint32_t nextMinLevelWindowLen{ 0 };
	std::vector<uint8_t> immediates;
	std::vector<uint8_t> mediums;
	std::vector<uint8_t> longs;
	std::vector<uint8_t> levels;
```

ibc commented on worker/src/RTC/ActiveSpeakerObserver.cpp:
```diff
@@ -0,0 +1,560 @@
+			MS_THROW_TYPE_ERROR("not an audio Producer");
+
+		this->mapProducerSpeaker[producer->id].producer = producer;
+		this->mapProducerSpeaker[producer->id].speaker  = new Speaker();
```
Here, if the same Producer is added twice, we are gonna leak the `speaker` class instance. This `AddProducer()` method should throw if the given Producer is already in the map.

NOTE: Yup, we don't check it in `AudioLevelObserver::AddProducer()`, but there nothing will leak.

ibc commented on worker/src/RTC/ActiveSpeakerObserver.cpp:
```diff
@@ -0,0 +1,560 @@
+		MS_TRACE();
+
+		uint64_t now              = DepLibUV::GetTimeMs();
+		uint64_t levelIdleTimeout = LevelIdleTimeout - (now - lastLevelIdleTime);
```
```suggestion
		uint64_t levelIdleTimeout = LevelIdleTimeout - (now - this->lastLevelIdleTime);
```

ibc commented on worker/src/RTC/ActiveSpeakerObserver.cpp:
```diff
@@ -0,0 +1,560 @@
+
+		if (levelIdleTimeout <= 0)
+		{
+			if (lastLevelIdleTime != 0)
```
```suggestion
			if (this->lastLevelIdleTime != 0)
```

ibc commented on worker/src/RTC/ActiveSpeakerObserver.cpp:
```diff
@@ -0,0 +1,560 @@
+			{
+				TimeoutIdleLevels(now);
+			}
+			lastLevelIdleTime = now;
```
```suggestion
			this->lastLevelIdleTime = now;
```

ibc commented on worker/src/RTC/ActiveSpeakerObserver.cpp:
```diff
@@ -0,0 +1,560 @@
+			lastLevelIdleTime = now;
+		}
+
+		if (mapProducerSpeaker.size() > 0 && CalculateActiveSpeaker())
```
```suggestion
		if (!this->mapProducerSpeaker.empty() && CalculateActiveSpeaker())
```

ibc commented on worker/src/RTC/ActiveSpeakerObserver.cpp:
```diff
@@ -0,0 +1,560 @@
+		MS_TRACE();
+
+		std::string newDominantId;
+		int32_t speakerCount = mapProducerSpeaker.size();
```
```suggestion
		int32_t speakerCount = this->mapProducerSpeaker.size();
```

ibc commented on worker/src/RTC/ActiveSpeakerObserver.cpp:
```diff
@@ -0,0 +1,560 @@
+		}
+		else if (speakerCount == 1)
+		{
+			auto it = mapProducerSpeaker.begin();
```
```suggestion
			auto it = this->mapProducerSpeaker.begin();
```

ibc commented on worker/src/RTC/ActiveSpeakerObserver.cpp:
```diff
@@ -0,0 +1,560 @@
+		else
+		{
+			Speaker* dominantSpeaker =
+			  (dominantId.empty()) ? nullptr : mapProducerSpeaker[dominantId].speaker;
```
```suggestion
			  (this->dominantId.empty()) ? nullptr : this->mapProducerSpeaker[this->dominantId].speaker;
```

ibc commented on worker/src/RTC/ActiveSpeakerObserver.cpp:
```diff
@@ -0,0 +1,560 @@
+			dominantSpeaker->EvalActivityScores();
+			double newDominantC2 = C2;
+
+			for (auto it = mapProducerSpeaker.begin(); it != mapProducerSpeaker.end(); ++it)
```
```suggestion
			for (auto it = this->mapProducerSpeaker.begin(); it != this->mapProducerSpeaker.end(); ++it)
```

ibc commented on worker/src/RTC/ActiveSpeakerObserver.cpp:
```diff
@@ -0,0 +1,560 @@
+
+			for (auto it = mapProducerSpeaker.begin(); it != mapProducerSpeaker.end(); ++it)
+			{
+				Speaker* speaker     = it->second.speaker;
+				const std::string id = it->second.producer->id;
```
Let's avoid string copy:

```suggestion
				Speaker* speaker      = it->second.speaker;
				const std::string& id  = it->second.producer->id;
```

ibc commented on worker/src/RTC/ActiveSpeakerObserver.cpp:
```diff
@@ -0,0 +1,560 @@
+
+				speaker->EvalActivityScores();
+
+				for (int interval = 0; interval < relativeSpeachActivitiesLen; interval++)
```
```suggestion
				for (int interval = 0; interval < this->relativeSpeachActivitiesLen; ++interval)
```

ibc commented on worker/src/RTC/ActiveSpeakerObserver.cpp:
```diff
@@ -0,0 +1,560 @@
+
+				for (int interval = 0; interval < relativeSpeachActivitiesLen; interval++)
+				{
+					relativeSpeachActivities[interval] = std::log(
```
```suggestion
					this->relativeSpeachActivities[interval] = std::log(
```

ibc commented on worker/src/RTC/ActiveSpeakerObserver.cpp:
```diff
@@ -0,0 +1,560 @@
+					  dominantSpeaker->GetActivityScore(interval) / speaker->GetActivityScore(interval));
+				}
+
+				double c1 = relativeSpeachActivities[0];
+				double c2 = relativeSpeachActivities[1];
+				double c3 = relativeSpeachActivities[2];
```
```suggestion
				double c1 = this->relativeSpeachActivities[0];
				double c2 = this->relativeSpeachActivities[1];
				double c3 = this->relativeSpeachActivities[2];
```

ibc commented on worker/src/RTC/ActiveSpeakerObserver.cpp:
```diff
@@ -0,0 +1,560 @@
+				Speaker* speaker     = it->second.speaker;
+				const std::string id = it->second.producer->id;
+
+				if (id == dominantId || speaker->paused)
```
```suggestion
				if (id == this->dominantId || speaker->paused)
```

ibc commented on worker/src/RTC/ActiveSpeakerObserver.cpp:
```diff
@@ -0,0 +1,560 @@
+			}
+		}
+
+		if (!newDominantId.empty() && newDominantId != dominantId)
+		{
+			dominantId = newDominantId;
```
```suggestion
		if (!newDominantId.empty() && newDominantId != this->dominantId)
		{
			this->dominantId = newDominantId;
```

ibc commented on worker/src/RTC/ActiveSpeakerObserver.cpp:
```diff
@@ -0,0 +1,560 @@
+	{
+		MS_TRACE();
+
+		for (auto it = mapProducerSpeaker.begin(); it != mapProducerSpeaker.end(); ++it)
+		{
+			Speaker* speaker      = it->second.speaker;
+			const std::string& id = it->second.producer->id;
+			uint64_t idle         = now - speaker->lastLevelChangeTime;
+
+			if (SpeakerIdleTimeout < idle && (dominantId.empty() || id != dominantId))
```
```suggestion
		for (auto it = this->mapProducerSpeaker.begin(); it != this->mapProducerSpeaker.end(); ++it)
		{
			Speaker* speaker      = it->second.speaker;
			const std::string& id = it->second.producer->id;
			uint64_t idle         = now - speaker->lastLevelChangeTime;

			if (SpeakerIdleTimeout < idle && (this->dominantId.empty() || id != this->dominantId))
```

ibc commented on worker/src/RTC/ActiveSpeakerObserver.cpp:
```diff
@@ -0,0 +1,560 @@
+	{
+		MS_TRACE();
+
+		minLevel               = MinLevel;
+		nextMinLevel           = MinLevel;
+		immediateActivityScore = MinActivityScore;
+		mediumActivityScore    = MinActivityScore;
+		longActivityScore      = MinActivityScore;
+
+		immediates.resize(ImmediateBuffLen);
+		mediums.resize(MediumsBuffLen);
+		longs.resize(LongsBuffLen);
+		levels.resize(LevelsBuffLen);
+
+		lastLevelChangeTime = DepLibUV::GetTimeMs();
```
```suggestion
		this->minLevel               = MinLevel;
		this->nextMinLevel           = MinLevel;
		this->immediateActivityScore = MinActivityScore;
		this->mediumActivityScore    = MinActivityScore;
		this->longActivityScore      = MinActivityScore;

		this->immediates.resize(ImmediateBuffLen);
		this->mediums.resize(MediumsBuffLen);
		this->longs.resize(LongsBuffLen);
		this->levels.resize(LevelsBuffLen);

		this->lastLevelChangeTime = DepLibUV::GetTimeMs();
```

ibc commented on worker/src/RTC/ActiveSpeakerObserver.cpp:
```diff
@@ -0,0 +1,560 @@
+
+		switch (interval)
+		{
+			case 0:
+				return immediateActivityScore;
+			case 1:
+				return mediumActivityScore;
+			case 2:
+				return longActivityScore;
```
```suggestion
			case 0:
				return this->immediateActivityScore;
			case 1:
				return this->mediumActivityScore;
			case 2:
				return this->longActivityScore;
```

ibc commented on worker/src/RTC/ActiveSpeakerObserver.cpp:
```diff
@@ -0,0 +1,560 @@
+
+	void ActiveSpeakerObserver::Speaker::LevelChanged(uint32_t level, uint64_t time)
+	{
+		if (lastLevelChangeTime <= time)
+		{
+			lastLevelChangeTime = time;
```
```suggestion
		if (this->lastLevelChangeTime <= time)
		{
			this->lastLevelChangeTime = time;
```

ibc commented on worker/src/RTC/ActiveSpeakerObserver.cpp:
```diff
@@ -0,0 +1,560 @@
+				b = level;
+			}
+
+			std::copy(levels.begin(), levels.end() - 1, levels.begin() + 1);
```
```suggestion
			std::copy(this->levels.begin(), this->levels.end() - 1, this->levels.begin() + 1);
```

ibc commented on worker/src/RTC/ActiveSpeakerObserver.cpp:
```diff
@@ -0,0 +1,560 @@
+
+			std::copy(levels.begin(), levels.end() - 1, levels.begin() + 1);
+
+			levels[0] = b;
```
```suggestion
			this->levels[0] = b;
```

ibc commented on worker/src/RTC/ActiveSpeakerObserver.cpp:
```diff
@@ -0,0 +1,560 @@
+	{
+		MS_TRACE();
+
+		LevelChanged(MinLevel, lastLevelChangeTime);
```
```suggestion
		LevelChanged(MinLevel, this->lastLevelChangeTime);
```

ibc commented on worker/src/RTC/ActiveSpeakerObserver.cpp:
```diff
@@ -0,0 +1,560 @@
+
+		for (uint32_t i = 0; i < ImmediateBuffLen; ++i)
+		{
+			uint8_t level = levels[i];
```
```suggestion
			uint8_t level = this->levels[i];
```

ibc commented on worker/src/RTC/ActiveSpeakerObserver.cpp:
```diff
@@ -0,0 +1,560 @@
+
+			uint8_t immediate = (level / SubunitLengthN1);
+
+			if (immediates[i] != immediate)
+			{
+				immediates[i] = immediate;
```
```suggestion
			if (this->immediates[i] != immediate)
			{
				this->immediates[i] = immediate;
```

ibc commented on worker/src/RTC/ActiveSpeakerObserver.cpp:
```diff
@@ -0,0 +1,560 @@
+	{
+		MS_TRACE();
+
+		return ComputeBigs(immediates, mediums, MediumThreshold);
```
```suggestion
		return ComputeBigs(this->immediates, this->mediums, MediumThreshold);
```

ibc commented on worker/src/RTC/ActiveSpeakerObserver.cpp:
```diff
@@ -0,0 +1,560 @@
+	{
+		MS_TRACE();
+
+		return ComputeBigs(mediums, longs, LongThreashold);
```
```suggestion
		return ComputeBigs(this->mediums, this->longs, LongThreashold);
```

ibc commented on worker/src/RTC/ActiveSpeakerObserver.cpp:
```diff
@@ -0,0 +1,560 @@
+	{
+		MS_TRACE();
+
+		immediateActivityScore = ComputeActivityScore(immediates[0], N1, 0.5, 0.78);
```
```suggestion
		this->immediateActivityScore = ComputeActivityScore(this->immediates[0], N1, 0.5, 0.78);
```

ibc commented on worker/src/RTC/ActiveSpeakerObserver.cpp:
```diff
@@ -0,0 +1,560 @@
+	{
+		MS_TRACE();
+
+		mediumActivityScore = ComputeActivityScore(mediums[0], N2, 0.5, 24);
+	}
+
+	void ActiveSpeakerObserver::Speaker::EvalLongActivityScore()
+	{
+		MS_TRACE();
+
+		longActivityScore = ComputeActivityScore(longs[0], N3, 0.5, 47);
```
```suggestion
		this->mediumActivityScore = ComputeActivityScore(this->mediums[0], N2, 0.5, 24);
	}

	void ActiveSpeakerObserver::Speaker::EvalLongActivityScore()
	{
		MS_TRACE();

		this->longActivityScore = ComputeActivityScore(this->longs[0], N3, 0.5, 47);
```

ibc commented on worker/src/RTC/ActiveSpeakerObserver.cpp:
```diff
@@ -0,0 +1,560 @@
+			return;
+		}
+
+		if ((minLevel == MinLevel) || (minLevel > level))
+		{
+			minLevel              = level;
+			nextMinLevel          = MinLevel;
+			nextMinLevelWindowLen = 0;
+		}
```
```suggestion
		if ((this->minLevel == MinLevel) || (this->minLevel > level))
		{
			this->minLevel              = level;
			this->nextMinLevel          = MinLevel;
			this->nextMinLevelWindowLen = 0;
		}
```

ibc commented on worker/src/RTC/ActiveSpeakerObserver.cpp:
```diff
@@ -0,0 +1,560 @@
+		}
+		else
+		{
+			if (nextMinLevel == MinLevel)
+			{
+				nextMinLevel          = level;
+				nextMinLevelWindowLen = 1;
+			}
+			else
+			{
+				if (nextMinLevel > level)
+				{
+					nextMinLevel = level;
+				}
+				nextMinLevelWindowLen++;
+				if (nextMinLevelWindowLen >= MinLevelWindowLen)
+				{
+					double newMinLevel = std::sqrt((double)(minLevel * nextMinLevel));
+
+					if (newMinLevel < MinLevel)
+					{
+						newMinLevel = MinLevel;
+					}
+					else if (newMinLevel > MaxLevel)
+					{
+						newMinLevel = MaxLevel;
+					}
+
+					minLevel = (int8_t)newMinLevel;
+
+					nextMinLevel          = MinLevel;
+					nextMinLevelWindowLen = 0;
+				}
```
```suggestion
			if (this->nextMinLevel == MinLevel)
			{
				this->nextMinLevel          = level;
				this->nextMinLevelWindowLen = 1;
			}
			else
			{
				if (this->nextMinLevel > level)
				{
					this->nextMinLevel = level;
				}
				this->nextMinLevelWindowLen++;

				if (this->nextMinLevelWindowLen >= MinLevelWindowLen)
				{
					double newMinLevel = std::sqrt(static_cast<double>(this->minLevel * this->nextMinLevel));

					if (this->newMinLevel < MinLevel)
					{
						this->newMinLevel = MinLevel;
					}
					else if (this->newMinLevel > MaxLevel)
					{
						this->newMinLevel = MaxLevel;
					}

					this->minLevel = static_cast<int8_t>(this->newMinLevel);

					this->nextMinLevel          = MinLevel;
					this->nextMinLevelWindowLen = 0;
				}
```

Commit:
Apply suggestions from code review

Co-authored-by: Iñaki Baz Castillo <ibc@aliax.net>

Commit:
Apply suggestions from code review.

ibc commented on worker/src/RTC/ActiveSpeakerObserver.cpp:
```diff
@@ -0,0 +1,564 @@
+		uint64_t now              = DepLibUV::GetTimeMs();
+		uint64_t levelIdleTimeout = LevelIdleTimeout - (now - this->lastLevelIdleTime);
+
+		if (levelIdleTimeout <= 0)
```
Codady says that this condition doesn't make sense since `levelIdleTimeout` is an unsigned number:

https://app.codacy.com/manual/versatica/mediasoup/pullRequest?bid=24405941&prid=7734463&utm_campaign=new_commit&utm_medium=Email&utm_source=Internal

So, I see two choices here (but I don't know which one is the correct one):

1. Make `levelIdleTimeout` be `int64_t`.
2. Or remove the condition.

SteveMcFarlin replied:
Given there is a subtraction involved in the calculation, I am going to error on the side of caution, and make this signed.

ibc commented on worker/src/RTC/ActiveSpeakerObserver.cpp:
```diff
@@ -0,0 +1,564 @@
+
+			if (dominantSpeaker == nullptr)
+			{
+				auto item       = mapProducerSpeaker.begin();
```
```suggestion
				auto item       = this->mapProducerSpeaker.begin();
```

please, ensure that all class members (excluding class methods) use `this->`.

Commit:
Change to signed integer.

Commit:
Apply code suggestion from code review

Co-authored-by: Iñaki Baz Castillo <ibc@aliax.net>

ibc commented:
Thanks, looking great. Waiting for CI to complete.

BTW could you please contribute to the `mediasoup-website` project with a PR to document the new API?

Commit:
Format

Commit:
Merge branch 'v3-dominant-speaker' of github.com:SteveMcFarlin/mediasoup into v3-dominant-speaker

ggarber committed:
rust fmt

ibc commented:
Thanks a lot for this, guys. I'll wait for the documentation PR before merging. Just let me know if you have any problem or question writing it. There is "gulp live" task in mediasoup-website project to run the web locally.

ibc commented:
Released in mediasoup v3.8.0 :)

Thanks a lot!

SteveMcFarlin commented:
Thank you for the great review process!

msach22 commented:
@SteveMcFarlin nice contribution 😄

mk19831002 commented:
Hi,
did you write the documentation for that, I didn't find it in API documentation.

ibc commented:
Somehow the website is not up to date in the web server, but yes, he wrote docs. I'll check and republish later today.

d2e1 commented:
Hello guys, 
I was testing the dominant speaker feature and a situation I came across prompted me to review the code. I also reviewed the Jitsi code and the article the algorithm is based on. As far as I can see, the following piece of code is added differently in Jitsi:

Mediasoup code:
https://github.com/versatica/mediasoup/blob/v3/worker/src/RTC/ActiveSpeakerObserver.cpp#L313

Jitsi code:
https://github.com/jitsi/jitsi-utils/blob/master/src/main/java/org/jitsi/utils/dsi/DominantSpeakerIdentification.java#L715

Related part of the article (Dominant Speaker Identification for Multipoint Videoconferencing by Ilana Volfin and Israel Cohen)
![image](https://user-images.githubusercontent.com/30109375/131319362-a6c70782-a102-4732-8dda-b97bd3da6491.png)


In the article and Jitsi code, the dominant speaker is included in the dividing part. I don't know the other parts of the code well, I wonder if a change is needed here?

Link for the  open discussion : https://mediasoup.discourse.group/t/question-regarding-active-speaker-observer-implementation/3227

punarinta commented:
Hi @SteveMcFarlin,

Could you please advise me what is the reason I'm not getting any events from ActiveSpeakerObserver instance?

I've initialized the top-level code like 
```js
const observer = await router.createActiveSpeakerObserver({interval: 500})

observer.on('dominantspeaker', (payload) => {
	console.log('event payload', payload)
})
```

For debugging I've added logs after lines 19 and 20 in https://github.com/versatica/mediasoup/blob/v3/node/lib/ActiveSpeakerObserver.js and I see that handleWorkerNotifications is called, but this.channel.on never fires its callback.

Both audio and video is sent and received correctly between the participating peers, so audio data definitely flows the way it should.

UPD: discussion moved to https://mediasoup.discourse.group/t/activespeakerobserver-does-not-emit-events/3621

ibc commented:
Please let use the forum for questions.
